### PR TITLE
WIP Quarantine gracefully downed node after some time

### PR DESCRIPTION
**Pull request by johanandren**
Refs #21521

### CODE REVIEW

File: akka-cluster/src/main/scala/akka/cluster/ClusterRemoteWatcher.scala [hunk 1]
```diff
@@ -90,14 +94,22 @@ private[cluster] class ClusterRemoteWatcher(
...
-      // The reason we don't quarantine gracefully removed members (leaving) is that
-      // Cluster Singleton need to exchange TakeOver/HandOver messages.
+
       if (previousStatus == MemberStatus.Down) {
         quarantine(m.address, Some(m.uniqueAddress.uid), s"Cluster member removed, previous status [$previousStatus]")
+      } else if (arteryEnabled) {
+        // don't quarantine gracefully removed members (leaving) directly,
+        // give Cluster Singleton some time to exchange TakeOver/HandOver messages.
+        import context.dispatcher
+        context.system.scheduler.scheduleOnce(cluster.downingProvider.downRemovalMargin, self, DelayedQuarantine(m, previousStatus))
```
Comment (johanandren) [inline akka-cluster/src/main/scala/akka/cluster/ClusterRemoteWatcher.scala]: Not sure this is a good idea, but if the cluster singleton doesn't have a specific removalMargin configured, this is what is used as removalMargin in ClusterSingletonManager.

I'm guessing we can't really depend on the singleton manager setting here.

**Comment by johanandren** [comment]
Any hints of how I could provide test coverage for this would be great.

File: akka-cluster/src/main/scala/akka/cluster/ClusterRemoteWatcher.scala (hunk 1 above)
Comment (patriknw) [inline akka-cluster/src/main/scala/akka/cluster/ClusterRemoteWatcher.scala]: I have created a new issue for redesigning the cluster leaving process, but that's a rather big thing. Until then we need something simple that can work, and I think a simple delay like this should be enough for now.

We already have enough test coverage, singleton tests will fail.

We can't use `downRemovalMargin`. It's by default zero. We should use something like 20 - 30 seconds. It doesn't feel great to introduce a new config property for this, but hardcoded value is not so nice either.

File: akka-cluster/src/main/resources/reference.conf [hunk 2]
```diff
@@ -54,6 +54,11 @@ akka {
     # `akka.cluster.DowningProvider` having a public one argument constructor accepting an `ActorSystem`
     downing-provider-class = ""
 
+    # Artery only setting
+    # When a node has been gracefully downed, let this time pass (to allow for example
```
Comment (patriknw) [inline akka-cluster/src/main/resources/reference.conf]: has been gracefully *removed*... and then quarantine the *removed* node.

File: akka-cluster/src/main/resources/reference.conf [hunk 3]
```diff
@@ -54,6 +54,11 @@ akka {
...
+    # cluster singleton handover to complete) and then quarantine the downed node.
+    quarantine-downed-node-after=30s
```
Comment (patriknw) [inline akka-cluster/src/main/resources/reference.conf]: `quarantine-removed-node-after`

downing is for the unexpected crash scenario, leaving is for the graceful removal

**Review by patriknw — CHANGES_REQUESTED** [review]
minor rename, but otherwise it looks good

File: akka-cluster/src/main/resources/reference.conf (hunk 3 above)
Comment (johanandren) [inline akka-cluster/src/main/resources/reference.conf]: D'oh!

**Review by patriknw — APPROVED** [review]

**Comment by patriknw** [comment]
LGTM

**Pull request merged**